tests/integration/test_charm.py

```diff
@@ -2,72 +2,116 @@
-    """
     charm = await ops_test.build_charm(".")
-    resources = {"oci-image": METADATA["resources"]["oci-image"]["upstream-source"]}
     await ops_test.model.deploy(
-        charm, resources=resources, application_name=APP_NAME, trust=True, series="jammy"
+        entity_url=str(charm),
+        resources={"oci-image": LOGIN_UI_IMAGE},
+        application_name=LOGIN_UI_APP,
+        trust=True,
+        series="jammy",
     )
 
-    async with ops_test.fast_forward():
-        await ops_test.model.wait_for_idle(
-            apps=[APP_NAME],
-            status="active",
-            raise_on_blocked=True,
-            timeout=1000,
-        )
-        assert ops_test.model.applications[APP_NAME].units[0].workload_status == "active"
-
-
-async def test_ingress_relation(ops_test: OpsTest):
     await ops_test.model.deploy(
-        TRAEFIK,
-        application_name=TRAEFIK_PUBLIC_APP,
+        ISTIO_INGRESS_CHARM,
+        application_name=INGRESS_APP,
+        trust=True,
         channel="latest/edge",
-        config={"external_hostname": "some_hostname"},
+        config={"external_hostname": PUBLIC_INGRESS_DOMAIN},
+    )
+
+    await ops_test.model.deploy(
+        CA_APP,
+        channel="latest/stable",
+        trust=True,
     )
 
-    await ops_test.model.add_relation(f"{APP_NAME}:ingress", TRAEFIK_PUBLIC_APP)
+    await ops_test.model.integrate(f"{INGRESS_APP}:certificates", f"{CA_APP}:certificates")
+    await ops_test.model.integrate(LOGIN_UI_APP, INGRESS_APP)
 
     await ops_test.model.wait_for_idle(
-        apps=[TRAEFIK_PUBLIC_APP],
+        apps=[LOGIN_UI_APP, INGRESS_APP, CA_APP],
         status="active",
         raise_on_blocked=True,
-        timeout=1000,
+        timeout=5 * 60,
     )
 
 
-async def test_has_ingress(ops_test: OpsTest):
-    # Get the traefik address and try to reach identity-platform-login-ui
-    public_address = await get_unit_address(ops_test, TRAEFIK_PUBLIC_APP, 0)
+async def test_ingress_integration(
+    ops_test: OpsTest,
+    leader_ingress_integration_data: Optional[dict],
+    public_ingress_address: str,
+    http_client: AsyncClient,
+) -> None:
+    assert leader_ingress_integration_data
+    assert leader_ingress_integration_data["ingress"]
 
-    resp = requests.get(f"http://{public_address}/{ops_test.model.name}-{APP_NAME}/ui/login")
+    data = json.loads(leader_ingress_integration_data["ingress"])
+    assert data["url"] == f"https://{PUBLIC_INGRESS_DOMAIN}/{ops_test.model_name}-{LOGIN_UI_APP}"
 
-    assert resp.status_code == 200
+    # Test HTTP to HTTPS redirection
+    http_endpoint = (
+        f"http://{public_ingress_address}/{ops_test.model_name}-{LOGIN_UI_APP}/ui/login"
+    )
+    resp = await http_client.get(
+        http_endpoint,
+        headers={"Host": PUBLIC_INGRESS_DOMAIN},
+    )
+    assert resp.status_code == http.HTTPStatus.MOVED_PERMANENTLY, (
+        f"Expected HTTP 301 for {http_endpoint}, got {resp.status_code}."
```

Comment:
Here a hardcoded value `301` is used. On the line before the same thing expressed via `http.HTTPStatus.MOVED_PERMANENTLY` constant. Does it make sense to stick to one of the options?

Reply:
IMHO, the `MOVED_PERMANENTLY` is for people who program the tests. The `301` in the assert failure message is for whoever runs the tests. So this way it should be readable for both.

Reply:
Oh, my bad. I understand what u mean. I can change it.